File: internal/auth/email_test.go
package auth

import "testing"

func TestValidateEmail(t *testing.T) {
	for _, tc := range []struct {
		in   string
		want string
		ok   bool
	}{
		{"admin@example.com", "admin@example.com", true},
		{"  admin@example.com ", "admin@example.com", true},
		{"first.last+tag@sub.example.co.uk", "first.last+tag@sub.example.co.uk", true},
		{"admin@[192.0.2.1]", "admin@[192.0.2.1]", true},
		{"admin@[2001:db8::1]", "admin@[2001:db8::1]", true},
		{"admin@[not-an-ip]", "", false},
		{"admin@[999.1.1.1]", "", false},
		{"admin@[IPv6:2001:db8::1]", "", false},
		{"admin@siloserver", "", false},
		{"admin@", "", false},
		{"@example.com", "", false},
		{"admin@example.", "", false},
		{"admin@.example", "", false},
		{"", "", false},
		{"   ", "", false},
		{"not an email", "", false},
		{"Admin <admin@example.com>", "", false},
		{"admin@example.com, other@example.com", "", false},
	} {
		got, err := ValidateEmail(tc.in)
		if (err == nil) != tc.ok {
			t.Errorf("ValidateEmail(%q) err = %v, want ok=%v", tc.in, err, tc.ok)
			continue
		}
		if got != tc.want {
			t.Errorf("ValidateEmail(%q) = %q, want %q", tc.in, got, tc.want)
		}
	}
}

File: web/src/lib/email.test.ts
import { describe, expect, it } from "vitest";

import { isValidEmail } from "./email";

describe("isValidEmail", () => {
  it.each([
    "admin@example.com",
    "  admin@example.com ",
    "first.last+tag@sub.example.co.uk",
    "admin@[192.0.2.1]",
    "admin@[2001:db8::1]",
  ])("accepts %s", (value) => {
    expect(isValidEmail(value)).toBe(true);
  });

  it.each([
    "admin@siloserver",
    "admin@",
    "@example.com",
    "admin@example.",
    "admin@.example",
    "",
    "   ",
    "not an email",
    "Admin <admin@example.com>",
    "a@b@example.com",
    "admin@[not-an-ip]",
    "admin@[999.1.1.1]",
    "admin@[IPv6:2001:db8::1]",
    "admin@[192.0.2.1",
  ])("rejects %s", (value) => {
    expect(isValidEmail(value)).toBe(false);
  });
});
